File: nim/bob/bob.nim
import std/strutils
import std/sequtils

proc hey*(s: string): string =
  var 
    message = s.strip()
    isQuestion = message.endsWith("?")
    isYelling = message.any(isAlphaAscii) and not message.any(isLowerAscii) 
  
  if message.len == 0:
    return "Fine. Be that way!"

  if isQuestion and isYelling:
      return "Calm down, I know what I'm doing!"

  if isQuestion:
      return "Sure."

  if isYelling:
      return "Whoa, chill out!"

  return "Whatever."

File: clojure/bob/src/bob.clj
(ns bob)

(require '[clojure.string :as str])

(defn has-char? [text]
  (boolean (re-find #"[a-zA-Z]" text)))

(defn yelling? [text]
  (and (has-char? text)
       (= (str/upper-case text) text)))

(defn asking? [text]
  (str/ends-with? text "?"))

(defn response-for [text]
  (let [s (str/trim text)
        ask? (asking? s)
        yell? (yelling? s)]

    (cond (empty? s) "Fine. Be that way!"
          (and ask? yell?) "Calm down, I know what I'm doing!"
          ask? "Sure."
          yell? "Whoa, chill out!"
          :else "Whatever.")))
